server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1098,6 +1102,122 @@ private void asyncCleanupUnlinkedShardLevelBlobs(
+            this.task = task;
+            this.shardIdentifier = indexShardIdentifier(indexUUID, shardId);
+            this.executor = executor;
+        }
+
+        private static String indexShardIdentifier(String indexUUID, ShardId shardId) {
+            return String.join("/", indexUUID, String.valueOf(shardId.id()));
+        }
+
+        @Override
+        public void run() {
+            if (ongoingRemoteDirectoryCleanups.add(shardIdentifier)) {
+                try {
+                    task.run();
+                    BlockingQueue<Runnable> pendingTasks = shardCleanupPendingTasks.get(shardIdentifier);
+                    try {
+                        if (pendingTasks != null) {
+                            for (Runnable pendingTask = pendingTasks.poll(0L, TimeUnit.MILLISECONDS); pendingTask != null;) {
+                                pendingTask.run();
+                            }
+                        }
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                } finally {
+                    ongoingRemoteDirectoryCleanups.remove(shardIdentifier);
+                }
+            } else {
+                shardCleanupPendingTasks.putIfAbsent(shardIdentifier, new LinkedBlockingQueue<>());
+                shardCleanupPendingTasks.get(shardIdentifier).add(task);
+            }
+        }
+    }
+
+    protected void releaseRemoteStoreLockAndCleanup(
+        String shardId,
+        String shallowSnapshotUUID,
+        BlobContainer shardContainer,
+        RemoteStoreLockManagerFactory remoteStoreLockManagerFactory
+    ) throws IOException {
+        if (remoteStoreLockManagerFactory == null) {
+            return;
+        }
+
+        RemoteStoreShardShallowCopySnapshot remoteStoreShardShallowCopySnapshot = REMOTE_STORE_SHARD_SHALLOW_COPY_SNAPSHOT_FORMAT.read(
+            shardContainer,
+            shallowSnapshotUUID,
+            namedXContentRegistry
+        );
+        String indexUUID = remoteStoreShardShallowCopySnapshot.getIndexUUID();
+        String remoteStoreRepoForIndex = remoteStoreShardShallowCopySnapshot.getRemoteStoreRepository();
+        // Releasing lock file before deleting the shallow-snap-UUID file because in case of any failure while
+        // releasing the lock file, we would still have the shallow-snap-UUID file and that would be used during
+        // next delete operation for releasing this lock file
+        RemoteStoreLockManager remoteStoreMetadataLockManager = remoteStoreLockManagerFactory.newLockManager(
+            remoteStoreRepoForIndex,
+            indexUUID,
+            shardId
+        );
+        remoteStoreMetadataLockManager.release(FileLockInfo.getLockInfoBuilder().withAcquirerId(shallowSnapshotUUID).build());
+        logger.debug("Successfully released lock for shard {} of index with uuid {}", shardId, indexUUID);
+        if (!isIndexPresent(clusterService, indexUUID)) {
```

Comment:
Should we have the check that if there are no locks then we can delete the entire directory in one go? Alternatively, if we are deleting the data incrementally without additional cost of GETs, we can go ahead with this option.

Reply:
we are deleting the data incrementally with updated `deleteStaleSegments` implementation. also doing it in one go could introduce other failure issues since we will end up deleting a lot of data.